R/nested_ifelse_linter.R

```diff
@@ -0,0 +1,49 @@
+#' Block usage of nested ifelse() calls
+#'
+#' Calling `ifelse` in nested calls is problematic for two main reasons:
+#'   1. It can be hard to read -- mapping the code to the expected output
+#'      for such code can be a messy task/require a lot of mental bandwidth,
+#'      especially for code that nests more than once
+#'   2. It is inefficient -- `ifelse` can evaluate _all_ of its arguments at
+#'      both yes and no (see https://stackoverflow.com/q/16275149); this issue
+#'      is exacerbated for nested calls
+#'
+#' Users can instead rely on a more readable alternative modeled after SQL
+#'   CASE WHEN statements, such as `data.table::fcase` or `dplyr::case_when`,
+#'   or use a look-up-and-merge approach (build a mapping table between values
+#'   and outputs and merge this to the input).
+#'
+#' @evalRd rd_tags("nested_ifelse_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+nested_ifelse_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    xpath <- glue::glue("
+    //expr[expr[SYMBOL_FUNCTION_CALL[ {xp_text_in_table(ifelse_funs)} ]]]
+    /expr[expr[SYMBOL_FUNCTION_CALL[ {xp_text_in_table(ifelse_funs)} ]]]
+    ")
+
+    bad_expr <- xml2::xml_find_all(xml, xpath)
+
+    return(lapply(
+      bad_expr,
+      xml_nodes_to_lint,
+      source_file = source_file,
+      lint_message = function(expr) {
+        matched_call <- xml2::xml_text(xml2::xml_find_first(expr, "expr/SYMBOL_FUNCTION_CALL"))
+        lint_message <- sprintf("Don't use nested %s() calls; instead, try", matched_call)
+        paste(lint_message, "(1) data.table::fcase; (2) dplyr::case_when; or (3) merging to a lookup table.")
```

Comment:
(3) may be better "using a lookup table". If the lookup is just ok one column, `new[match(Col, old)]` in general or `unname(new[old])` for charactes is the optimal code in my testing.

Or maybe "merging or matching to a lookup table"?

Reply:
good idea. i like "using a lookup table". The merging approach is pretty ugly in general, if it's needed. The simple `named_vector[[lookup_keys]]` approach is nicer when it's possible.